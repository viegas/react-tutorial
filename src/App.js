import React, { Component, useState, useEffect } from 'react';
import './App.css';

const App = ({ titulo }) => {
    const [val, setVal] = useState(0);

    useEffect(() => {
        console.log('Mountou o functional component');
    }, []);

    const fn = () => setVal(val + 1);

    return (
        <div className="App">
            <span
                style={{
                    color: 'blue',
                    backgroundColor: 'yellow',
                }}
            >
                {' '}
                Meu todo {titulo}{' '}
            </span>
            <span> Meu todo {val} </span>
            <button onClick={fn}>btn</button>
        </div>
    );
};

// -----------------------------------

export class App2 extends Component {
    state = {
        test2: 0,
    };

    componentDidMount() {
        console.log(' Componente montou ');
    }

    exemploFn = (param) => {
        const { test2 } = this.state;
        this.setState({
            test2: test2 + 1,
        });
    };

    render() {
        const { test } = this.props;
        const { test2 } = this.state;

        return (
            <div className="App">
                <span> Meu todo {test}</span>
                <span> Meu todo {test2}</span>
                <button onClick={this.exemploFn}>btn</button>
            </div>
        );
    }
}

export default App;
