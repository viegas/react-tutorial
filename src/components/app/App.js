import React, { useState } from 'react';
import styles from './App.module.css';

import Title from '../title/Title';
import Input from '../input/Input';
import TaskList from '../taskList/TaskList';
import Footer from '../footer/Footer';
import Box from '../box/Box';

import useTodoHooks from './App.hooks';

const initValue = [{ id: 0, text: 'My first todo' }];

const App = () => {
    const [todoList, setTodoList] = useState(initValue);

    const { useCreateTodo, useUpdateTodo, useRemoveTodo } = useTodoHooks(
        todoList,
        setTodoList
    );

    return (
        <div className={styles.app}>
            <Title />
            <Box>
                <Input onCreateTodo={useCreateTodo} />
                <TaskList
                    list={todoList}
                    onRemoveItem={useRemoveTodo}
                    onUpdateItem={useUpdateTodo}
                />
                <Footer size={todoList.length} />
            </Box>
        </div>
    );
};

export default App;
