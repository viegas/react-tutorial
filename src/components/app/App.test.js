import App from './App';
import nock from 'nock';

import useTodoHooks from './App.hooks';
import { mount } from 'enzyme';
import { act } from 'react-test-renderer';

it('renders correctly on loading data', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
});

it('renders correctly with data', () => {
    // nock('http://localhost:8978/')
    //     .get('/todos')
    //     .reply(200, {
    //         data: [
    //             {
    //                 text: 'some todo',
    //                 id: 1,
    //             },
    //         ],
    //     });

    const wrapper = mount(<App />);
    act(() => {
        wrapper.update();
    });

    expect(wrapper).toMatchSnapshot();
});

it('Should find the intern components', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find('Title')).toHaveLength(1);
    expect(wrapper.find('Input')).toHaveLength(1);
    expect(wrapper.find('TaskList')).toHaveLength(1);
    expect(wrapper.find('Footer')).toHaveLength(1);
});

describe('App hooks test', () => {
    let setValue;

    beforeEach(() => {
        setValue = jest.fn();
    });

    it.only('Should call useCreateTodo correctly', async () => {
        const todoText = 'new todo text';

        nock('http://localhost:8978')
            .get('/todos')
            .reply(200, [
                {
                    text: 'some todo',
                    id: 1,
                },
            ]);

        nock('http://localhost:8978').post('/todos').reply(200);

        const { useCreateTodo } = useTodoHooks([], setValue);
        await useCreateTodo(todoText);

        expect(setValue).toHaveBeenCalledTimes(2);

        expect(setValue).toHaveNthReturnedWith(2, {});
    });

    it('Should call useUpdateTodo correctly', () => {
        const { useUpdateTodo } = useTodoHooks(
            [
                { id: 0, text: 'todo' },
                { id: 1, text: 'other' },
            ],
            setValue
        );

        useUpdateTodo(0, 'Updated');
        expect(setValue).toBeCalledWith([
            { id: 0, text: 'Updated' },
            { id: 1, text: 'other' },
        ]);
    });

    it('Should call useRemoveTodo correctly', () => {
        const { useRemoveTodo } = useTodoHooks(
            [{ id: 0, text: 'todo' }],
            setValue
        );
        useRemoveTodo(0);
        expect(setValue).toBeCalledWith([]);
    });
});
