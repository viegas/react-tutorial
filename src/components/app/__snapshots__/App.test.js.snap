// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly on loading data 1`] = `
<div
  className="app"
>
  <Title />
  <Box>
    <Input
      onCreateTodo={[Function]}
      onSelectAll={[Function]}
    />
    <span>
      LOADING....
    </span>
    <Footer
      size={0}
    />
  </Box>
</div>
`;

exports[`renders correctly with data 1`] = `
<App>
  <div
    className="app"
  >
    <Title>
      <h1
        className="header"
      >
        Todo App
      </h1>
    </Title>
    <Box>
      <div
        className="wrapper"
      >
        <Input
          onCreateTodo={[Function]}
          onSelectAll={[Function]}
        >
          <div
            className="wrapper"
          >
            <input
              onClick={[Function]}
              type="checkbox"
            />
            <input
              className="input"
              onKeyPress={[Function]}
              placeholder="What needs to be done?"
              type="text"
            />
          </div>
        </Input>
        <span>
          LOADING....
        </span>
        <Footer
          size={0}
        />
      </div>
    </Box>
  </div>
</App>
`;
